043-tasks_queue/proposal.md

```diff
@@ -0,0 +1,145 @@
+
+In this schema the resource pool only increases and decreases the number of active tasks on the workers.  In the next section we show that the proposal is not specific to the active tasks. The resource pool could be aware of other parameters that describe the workers' available capacity.
+
+## Resource Pool
+
+The new ResourcePool component has the following responsibilities:
+
+* observe the number of active tasks on the worker
+* offer workers' resources to tasks
+
+Concretely, we propose the following interface
+```go
+type Resources = map[string]int  // for example: {"active-task": 1}
+
+type ResourcePool interface {
+    RequestWorker(requiredResources Resources, strategy PlacementStrategy, task parameters...) -> Worker
+    ReleaseWorker(requiredResources Resources, worker Worker)
+}
+```
+
+This interface mimics the fundamental operations of locks and semaphores.
+
+Now, the interaction between the Client and the ResourcePool integrates unobtrusively into the current code base. Instead of tasks sending themselves immediately to the workers, the same part of the code could be, schematically
+
+```go
+func (client *client) RunTaskStep(...) TaskResult {
+    //...
+
+    requiredResources := make(map[string]int)
+    requiredResources["active-task"] = 1
+ 
+    chosenWorker := resourcePool.RequestWorker(requiredResoures, placementStrategy, taskParameters...)
+    defer resourcePool.ReleaseWorker(requiredREsources, chosenWorker)
+
+    //...now the worker can start the task container...
+}
+```
+
+## Implementation as a task queue
+
+This proposal doesn't want to implement a full-fledged workload scheduler into Concourse, but a simple mechanism by which the workers are protected from overload and each task gets a fair share of the workers' resources.
+
+In the pull request [#4612](https://github.com/concourse/concourse/pull/4612) a task queue was proposed. This task was rejected because the implementation was too specific for the `limit-active-task` placement strategy. We believe, however, that the components from that PR are still valuable. At least, it's a good starting point.
+
+The first resource pool implementation can be a simple task queue:
+
+* The RequestWorker messages are queued up. The queue is stored in the database
+* As resources become available they are offered to the first task in the queue.
+
+This resource pool implementation would be independent of the chosen worker placement strategy. Thus all worker placement strategies benefit from a safe and fair task distribution process.
+
+# Open Questions
+
+* Is "Resource Pool" a good name? (resource means something different in the context of pipeline jobs and tasks)
+* Could the functionality of the "Resource Pool"' be fit some already existing component?
+* Should this be tasks scheduling or more generic step scheduling? That is to include check containers (resource-type checks etc.)
+* What is the relationship to the new LIDAR component?
```

Comment:
There shouldn't be any more complexity with the new lidar component, it is just a different way of scheduling checks as the checks are queued and run as a step https://github.com/concourse/concourse/blob/master/atc/exec/check_step.go. If we wanted to later limit the number of checks run on a worker (not sure if we would ever want that though because checks should be incredibly quick to run), we could add the limiting functionality to the resource pool when we fetch the worker to run the check step on.